package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -645,51 +642,29 @@ cdef class _NSGrid(object):
 
         return True
 
-    cdef fill_grid(self, real[:, ::1] coords):
+    cdef void fill_grid(self, real[:, ::1] coords) nogil:
         """Sorts atoms into cells based on their position in the brick shaped box
 
-        Every atom inside the brick shaped box is assigned a
-        cell-id based on its position. Another list ``beadids``
-        sort the atom-ids in each cell.
-
         Note
         ----
         The method fails if any coordinate is outside the brick shaped box.
 
         """
+        cdef ns_int i, cellindex
 
-        cdef ns_int i, cellindex = -1
```

Comment:
The old data structure was a sort of 2 dimensional array, which required two passes through the data to first calculate the size to allocate, then another to fill this array.  Instead a simple linked list (of fixed and smaller size) can be used